### PR TITLE
[SOAR-11442] InsightVM New Scans  Trigger Fix

### PR DESCRIPTION
**Pull request by ablakley-r7**
### Description

Describe the proposed changes:

  - `TypeError: argument of type 'NoneType' is not iterable` exception raised during New Scans trigger
  - `output = func(params)\n  File \"/usr/local/lib/python3.7/site-packages/rapid7_insightvm_rapid7_plugin-4.9.1-py3.7.egg/komand_rapid7_insightvm/triggers/new_scans/trigger.py\", line 64, in run\n    if scan.get(\"scan_id\") in cache_site_scans.get(site_id):\nTypeError: argument of type 'NoneType' is not iterable\n", "status": "error"`

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py [hunk 1]
```diff
@@ -116,6 +98,36 @@ def scans_query(scan_statuses, site_ids):
...
+    def get_scan_details(self, endpoint: str, scan: dict):
+        response = self.connection.session.get(url=endpoint, verify=False)
+        if response.status_code in [200, 201]:  # 200 is documented, 201 is undocumented
+            try:
+                scan_details = response.json()
+            except json.decoder.JSONDecodeError as e:
+                raise PluginException(
+                    cause=f"Error: Failed to parse response while retrieving scan "
+                    f"details for scan ID {scan['scan_id']}.",
+                    assistance=f"Exception returned was {e}",
```
Comment (igorski-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py:122]:
```suggestion
            except json.decoder.JSONDecodeError as error:
                raise PluginException(
                    cause=f"Error: Failed to parse response while retrieving scan "
                    f"details for scan ID {scan['scan_id']}.",
                    assistance=f"Exception returned was {error}",
```

**Review by igorski-r7 — CHANGES_REQUESTED** [review]
In general, looks great! One small comment 😃

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py [hunk 2]
```diff
@@ -52,36 +50,18 @@ def run(self, params={}):
...
-                            )
-                    else:
-                        raise PluginException(
-                            cause=f"ERROR: Failed to retrieve scan ID {scan['scan_id']}.",
-                            assistance=f"Status code returned was {response.status_code}",
-                        )
+                    scan_details = NewScans.get_scan_details(self, endpoint, scan)
 
                     # Add site name and id; not provided by endpoint
                     scan_details["siteId"] = scan.get("site_id")
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py:67]:
```suggestion
                    scan_details["siteId"] = scan.get("site_id", {})
```

**Review by cmcnally-r7 — CHANGES_REQUESTED** [review]
Looks good! Some small comments 😃

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py [hunk 3]
```diff
@@ -52,36 +50,18 @@ def run(self, params={}):
...
-            except ValueError as e:
-                raise PluginException(cause="Failed to load cache file", assistance=f"Exception returned was {e}")
+            cache_site_scans = NewScans.get_cache_site_scans(self)
 
             # Send scans based on configuration
             for site_id, scans in site_scans.items():
                 for scan in scans:
                     # Only process scan IDs not previously cached
-                    if scan.get("scan_id") in cache_site_scans.get(site_id):
+                    if scan.get("scan_id") in cache_site_scans.get(site_id, []):
```
[the remark this answers is not in the document]
Comment (ablakley-r7) [reply plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py:59]: I think these are fine as they default value isn't used and is redundant.

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py [hunk 4]
```diff
@@ -92,6 +72,8 @@ def run(self, params={}):
                     self.send({Output.SCAN: scan_details})
 
                     # Update cache
+                    if cache_site_scans.get(site_id) is None:
```
[the remark this answers is not in the document]
Comment (ablakley-r7) [reply plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py:75]: I think we should leave it out as None is default and more true to the statement.

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py (hunk 2 above)
Comment (ablakley-r7) [reply plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py:67]: Default should be fine here as well as None or {} will be the same outcome.

**Review by igorski-r7 — APPROVED** [review]

**Review by cmcnally-r7 — APPROVED** [review]

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py [hunk 5]
```diff
@@ -116,6 +100,36 @@ def scans_query(scan_statuses, site_ids):
...
+                    f"details for scan ID {scan['scan_id']}.",
+                    assistance=f"Exception returned was {error}",
+                )
+        else:
+            raise PluginException(
+                cause=f"ERROR: Failed to retrieve scan ID {scan['scan_id']}.",
+                assistance=f"Status code returned was {response.status_code}",
+            )
+        return scan_details
+
```
Comment (jkaluza-rapid7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/new_scans/trigger.py:132]: Could you add unit-tests just for the new methods?

**Review by jkaluza-rapid7 — CHANGES_REQUESTED** [review]

**Review by jkaluza-rapid7 — APPROVED** [review]

**Review by jkaluza-rapid7 — APPROVED** [review]

**Pull request merged**